coins/monero/src/wallet/seed/classic/mod.rs

```diff
@@ -0,0 +1,250 @@
+  let word_list_length = words.len() as u32;
+
+  // To store the found words & add the checksum word later.
+  let mut result = vec![];
+
+  // convert to words
+  // 4 bytes -> 3 words. 8 digits base 16 -> 3 digits base 1626
+  for i in 0 .. 8 {
+    let mut indices: [u32; 4] = [0; 4];
+    // convert first 4 byte to u32 & get the word indices
+    let start = i * 4;
+    // convert 4 byte to u32
+    indices[0] = u32::from_le_bytes(key[start .. start + 4].try_into().unwrap());
+    indices[1] = indices[0] % word_list_length;
+    indices[2] = (indices[0] / word_list_length + indices[1]) % word_list_length;
+    indices[3] =
+      (((indices[0] / word_list_length) / word_list_length) + indices[2]) % word_list_length;
+
+    // append words to seed
+    result.push(words[indices[1] as usize].clone());
+    result.push(words[indices[2] as usize].clone());
+    result.push(words[indices[3] as usize].clone());
+  }
+  let checksum_index = create_checksum_index(&result, &lang)?;
+  result.push(result[checksum_index].clone());
+
+  Ok(result.join(" "))
+}
+
+/// takes a seed and returns the code for matching language
+fn find_seed_language(
+  seed: &[String],
+  has_checksum: bool,
+) -> Result<(Vec<usize>, Language), SeedError> {
+  let lang_list = languages::all()?;
+  let mut matched_indices = vec![];
+  let matched_language: Language;
+
+  for lang in &lang_list {
+    let word_map = &lang.word_map;
+    let trimmed_map = &lang.trimmed_word_map;
+
+    // Iterate through all the words and see if they're all present
+    let mut full_match = true;
+    for w in seed {
+      if has_checksum {
+        let trimmed_word = utf8_prefix(w, lang.unique_prefix_length);
+        if !trimmed_map.contains_key(&trimmed_word) {
+          full_match = false;
+          break;
+        }
+        matched_indices.push(trimmed_map[&trimmed_word]);
+      } else {
+        if !word_map.contains_key(w) {
+          full_match = false;
+          break;
+        }
+        matched_indices.push(word_map[w]);
+      }
+    }
+
+    if full_match && has_checksum && !checksum_test(seed, lang)? {
+      full_match = false;
+    }
+
+    if full_match {
+      matched_language = lang.clone();
+      return Ok((matched_indices, matched_language));
+    }
+
+    // no match for this language clear the indices
+    matched_indices.clear();
+  }
+
+  Err(SeedError::UnknownLanguage)?
+}
+
+fn utf8_prefix(word: &str, mut prefix_len: usize) -> String {
+  let mut prefix = String::new();
+  for a in word.chars() {
+    if (a as u8) & 0xc0 == 0x80 {
+      prefix.push(a);
+      continue;
+    }
+    prefix.push(a);
+    prefix_len -= 1;
+
+    if prefix_len == 0 {
+      break;
+    }
+  }
+  prefix
+}
+
+fn create_checksum_index(words: &[String], lang: &Language) -> Result<usize, SeedError> {
+  let mut trimmed_words = String::new();
+
+  for w in words {
+    let word = utf8_prefix(w, lang.unique_prefix_length);
+    if !lang.trimmed_word_map.contains_key(&word) {
```

Comment:
Earlier, an invalid trimmed_word_map was created. Is this supposed to use trimmed_word_map? I thought the intent, for languages supporting trimmed words, was that wallets would expand to full words and then use those.

Reply:
As far as I read and understand the code, using trimmed words isn't about the language, it is about the specific seed type. Namely, one with checksum or without. It always uses the trimmed words for seed with checksum, uses the full words if there is not checksum in the seed. And it always make a seed with checksum when creating a new random seed, so it almost always uses trimmed words. I think the only reason using full words is there for historic seeds that doesn't have a checksum. Or you can create a seed without a checksum by will. It is just to have as support I guess.

Reply:
Per BIP39, all words are distinct after some amount of characters. That enables recovery of a seed, even if you only have 4 chars from each word. I don't believe it's about being forced to use trimmed/untrimmed. There's just no specific benefit to internally loading untrimmed seeds.

My primary comment is about how this assumes trimmed_word_map is valid when it was placed in an invalid state earlier, and how this shouldn't work on languages not supporting trimming.

Reply:
We may need to
```rs
let mut word
if trimmed_word_map.is_some() {
  word = trimmed_word_map[word[.. unique_prefix_length]
}
```
?